File: Lex-Java/Lex.java
//Patrick Stumps
//pstumps@ucsc.edu
//CMPS 101 
//PA 1
//Sorts items from an array in alphabetical order
import java.util.Scanner;
import java.io.*;

public class Lex {
	public static void main(String[] args) throws IOException{
		
		if (args.length > 2 || args.length < 2) {
			System.err.println("Usage: Lex input filename output filename");
			System.exit(0);
        }
        
        int inputLines=0;
        String line = null;
        String input[]= null;
        List sorted = new List();
        sorted.append(0);
		
        Scanner sc = new Scanner(new File(args[0]));
        Scanner in = new Scanner(new File(args[0]));
        PrintWriter out = new PrintWriter(new FileWriter(args[1]));
		
		while (sc.hasNextLine()) {
            inputLines++;
            sc.nextLine();
        }

        input = new String[inputLines];

        for (int i=0; i<inputLines; i++){
            input[i]= in.nextLine();
        }

        //debugging
        //sorted.moveFront();
        //System.out.println(sorted.get());
        for (int i = 0; i < inputLines-1; i++) {
            sorted.moveFront();
            for (int j = -i ; j < 1; j++) {
                if (input[sorted.get()].compareTo(input[i+1])>0) {  
                    sorted.insertBefore(i+1);
                    j = 1;
                    //System.out.println("Comparing " +input[sorted.get()]+ " to " +input[i+1]);
                    //System.out.println(sorted);
                }
                else if (sorted.index() == sorted.length()-1) {
                    //System.out.println("Appending " +input[i+1]);
                    sorted.append(i+1);
                    j = 1;
                }		 
                else {
                    sorted.moveNext(); 		  
                }
            }  
        }
        //System.out.print(sorted);
        sorted.moveFront();
        //System.out.println();
        //System.out.println(sorted.get());
        while(sorted.index() != -1){
            out.println(input[sorted.get()]);
            sorted.moveNext();
        }

        sc.close();
        out.close();
		
    }
}

File: Lex-C/Lex.c
//Patrick Stumps
//pstumps@ucsc.edu
//CMPS101
//PA2
//Lex.c

#include<stdlib.h>
#include<stdio.h>
#include<string.h>
#include "List.h"

#define MAX_LEN 256

int main(int argc,  char*argv[]){
    if(argc != 3){
        printf("Usage: Lex inputfile outputfile");
        exit(1);
    }

    //FILEIO declaration
    FILE *input = fopen(argv[1], "r");
    FILE *output = fopen(argv[2], "w");

    int inputNum = 0, p = 0, l=0;
    char inputLine[MAX_LEN];
    List sList = newList();
    append(sList, 0);

    //Determine # of lines from input file
    //printf("Counting lines from input file \n");
    while(fgets(inputLine, MAX_LEN, input) != NULL){
        inputNum++;
    }
    
    //Creating array to be sorted
    char *unSorted[inputNum];
    rewind(input);

    //Allocating memory for sorted 
    //printf("Allocating memory for Array\n");
    for (int i=0; i <= inputNum; i++){
        unSorted[i] = calloc(1, sizeof(char *[MAX_LEN]));
    }

    //Copying input from inputLine to array
    //printf("Copying lines from input file to array\n");

    while(fgets(inputLine, MAX_LEN, input) != NULL){
        strcpy(unSorted[p], inputLine);
        p++;
    }

    //Sorting input
    //printf("Sorting...\n");
    for (int j=0; j<inputNum-1; j++){
        moveFront(sList);
        for(int k=-j; k<1; k++){
            //printlist(stdout, sList);
            if(strcmp(unSorted[get(sList)], unSorted[j+1]) > 0){
                //printf("Comparing %s to %s\n", unSorted[get(sList)], unSorted[j+1]);
                insertBefore(sList, j+1);
                k=1;
            } 
            else if (index(sList) == length(sList)-1){
                //printf("Appending %d to list\n", j+1);
                append(sList, j+1);
                k=1;
            }
            else{
                //printf("Using MoveNext");
                moveNext(sList);
            }
        }
    }

    moveFront(sList);

    //Printing sorted data to output file
    //prinf("Printing to output file \n");
    while(index(sList) != -1){
        fprintf(output, "%s", unSorted[get(sList)]);
        moveNext(sList);
    }

    while(l <= inputNum){
        free(unSorted[l]);
        l++;
    }
    
    freeList(&sList);
}